File: data_structures/python/SinglyCircularLL.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class SinglyCircularLL:
    def __init__(self):
        self.head = None
        self.tail = None

    def InsertFirst(self,data):
        new_node = Node(data)
        if self.head == None and self.tail == None:#LL is empty
            self.head = new_node
            self.tail = new_node
            new_node.next = self.head

        else:#LL contains atleast one node
            new_node.next = self.head
            self.head = new_node
            self.tail.next = self.head

    def InsertLast(self,data):
        new_node = Node(data)
        temp = self.head
        if self.head == None and self.tail == None:#LL is empty
            self.head = new_node
            self.tail = new_node
            new_node.next = self.head

        else:#LL contains atleast one node
            self.tail.next = new_node
            new_node.next = self.head
            self.tail = new_node

    def DeleteFirst(self):
        if self.head == None and self.tail == None:#LL is empty
            print("Linked list is empty")
            return
        
        if self.head == self.tail:#LL contains only one node
            self.head = None
            self.tail = None

        else:#LL contains atleast one node
            self.head = self.head.next
            self.tail.next = self.head

    def DeleteLast(self):
        if self.head == None and self.tail == None:#LL is empty
            print("Linked list is empty")
            return
        
        if self.head == self.tail:#LL contains only one node
            self.head = None
            self.tail = None

        else:#LL contains atleast one node
            temp = self.head
            while(temp.next != self.tail):
                temp = temp.next
            temp.next = self.head
            self.tail = temp
            
    def Display(self):
        if(self.head == None):
            print("Linked List is empty")
            return
        
        temp = self.head
        print("<-", end="")
        while True:
                print(f"| {temp.data} |->", end="")

                if temp == self.tail: 
                    break 
                temp = temp.next                
        print()

    def Count(self):
        iCount = 0
        if(self.head == None):
            return 0
        
        temp = self.head

        while True:
            iCount = iCount + 1

            if(temp == self.tail):
                break
            temp = temp.next
        return iCount
    
    def InsertAtPos(self,data,pos):
                
        new_node = Node(data)

        temp = self.head

        iCount = self.Count()

        if(pos < 1 or pos > iCount + 1):
            print("Invalid position")
            return
        
        if(pos == 1):
            self.InsertFirst(data)

        elif(pos == iCount + 1):
            self.InsertLast(data)

        else:
            for i in range(1,pos-1):
                temp = temp.next

            new_node.next = temp.next
            temp.next = new_node

    def DeleteAtPos(self,pos):

        temp = self.head

        iCount = self.Count()

        if(pos < 1 or pos > iCount):
            print("Invalid position")
            return
        
        if(pos == 1):
            self.DeleteFirst()

        elif(pos == iCount ):
            self.DeleteLast()

        else:
            temp = self.head
            for i in range(1,pos-1):
                temp = temp.next

            target = temp.next
            temp.next = target.next

def main():
    sobj = SinglyCircularLL()

    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    sobj.InsertLast(101)
    sobj.InsertLast(111)
    sobj.InsertLast(121)

    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

    sobj.InsertAtPos(105,5)
    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

    sobj.DeleteAtPos(5)
    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

if __name__ == "__main__":
    main()        
        

File: data_structures/java/SinglyCircularLL.java
class node
{
    public int data;
    public node next;
}

class SinglyCircularLL
{
    private node first;
    private node last;
    private int iCount;

    public SinglyCircularLL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;

        if(first == null && last == null)//LL is empty
        {
            first = newn;
            last = newn;

            last.next = first;
        }
        else 
        {
            newn.next = first;

            first = newn;
            last.next = first;
        }
        iCount++;
    }

    public void InsertLast(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;


        if(first == null && last == null)//ll is empty
        {
            first = newn;
            last = newn;
            last.next = first;
        }
        else
        {
            last.next = newn;
            last = newn;
            last.next = first;
        }
        iCount++;
    }

    public void Display()
    {   
        node temp = null;

        if(first == null && last == null)
        {
            return;
        }
        temp = first;
        do 
        {
            System.out.print("|"+ temp.data + " | ->");
            temp = temp.next;
        }
        while(temp != first);
        System.out.println();

        
    }

    public int Count()
    {
        return iCount;
    }

    public void InsertAtPos(int no,int pos)
    {

        
        if(pos < 1 || pos > iCount + 1)
        {
            System.out.println("Invalid position");
            return;
        }
        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            node temp = null;

            node newn = null;
            newn = new node();

            newn.data = no;
            newn.next = null;

            int iCnt = 0;

            temp = first;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next = newn;
            iCount++;
        }
        

    }

    public void DeleteFirst()
    {
        node temp = null;

        if(first == null && last == null)//LL is empty
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            temp = first;
            first = temp.next;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        node temp = null;

        if(first == null && last == null)
        {
            return;
        }
        else if(first.next == first)
        {
            first = null;
            last = null;
        }
        else
        {
            temp = first;
            while(temp.next != last)
            {
                temp = temp.next;
            }
            last = temp;
            temp.next = first;

        }
        iCount--;
    }

    public void DeleteAtPos(int pos)
    {   


        if(pos < 1 || pos > iCount)
        {
            System.out.println("Invalid position");
            return;
        }
        else if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount)
        {
            DeleteLast();
        }
        else
        {   
            int iCnt = 0;
            node temp = null;
            node target = null;
            temp = first;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp.next;
            }
            target = temp.next;

            temp.next = target.next;

            iCount--;
        }
    }
}

class program377
{
 public static void main(String A[])
    {
        SinglyCircularLL sobj = new SinglyCircularLL();

        int iRet = 0;

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);


        sobj.InsertAtPos(105,5);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);

        sobj.DeleteAtPos(5);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);
    }
}
